File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        int count = 1;
        int amount = 0;
        for (ListNode* p = head; p; p = p->next) amount++;
        
        ListNode* p = head;
        while (count <= amount - n)
        {
            p = p->next;
            count++;
        }
        if (p == head)
        {
            head = p->next;
            delete p;
            return head;
        }
        ListNode* t = head;
        while (t->next != p) t = t->next;
        t->next = p->next;
        p->next = nullptr;
        delete p;
        return head;
    }
};

File: 19-remove-nth-node-from-end-of-list/remove-nth-node-from-end-of-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode removeNthFromEnd(ListNode head, int n) {
        int count = 0;
        for (ListNode tmp = head; tmp != null; tmp = tmp.next) {
            count++;
        }

        ListNode dummy = new ListNode(0);
        dummy.next = head;

        if (count == n) {
            dummy.next = head.next;
            return dummy.next;
        }

        int ith = 0;
        for (ListNode tmp = head; tmp != null; tmp = tmp.next) {
            ith++;

            if (ith == count - n) {
                tmp.next = tmp.next.next;
                break;
            }
        }

        return head;
    }
}